src/jit/earlyprop.cpp

```diff
@@ -307,72 +309,57 @@ bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
             {
                 GenTreeBoundsChk* check = tree->gtNext->AsBoundsChk();
 
-                if ((check->gtArrLen == tree) && check->gtIndex->IsCnsIntOrI() &&
-                    (check->gtIndex->AsIntCon()->IconValue() >= 0) &&
-                    (check->gtIndex->AsIntCon()->IconValue() < actualVal->AsIntCon()->IconValue()))
+                if ((check->gtArrLen == tree) && check->gtIndex->IsCnsIntOrI())
                 {
-                    GenTree* comma = check->gtGetParent(nullptr);
-
-                    if ((comma != nullptr) && comma->OperIs(GT_COMMA) && (comma->gtGetOp1() == check))
+                    ssize_t checkConstVal = check->gtIndex->AsIntCon()->IconValue();
+                    if ((checkConstVal >= 0) && (checkConstVal < actualConstVal))
                     {
-                        GenTree* next = check->gtNext;
-                        optRemoveRangeCheck(comma, root);
-                        // Both `tree` and `check` have been removed from the statement. Ensure that optEarlyProp
-                        // can process the rest of the statment by changing tree->gtNext appropriately.
-                        tree->gtNext = next;
-                        return true;
+                        GenTree* comma = check->gtGetParent(nullptr);
+                        if ((comma != nullptr) && comma->OperIs(GT_COMMA) && (comma->gtGetOp1() == check))
+                        {
+                            GenTree* next = check->gtNext;
+                            optRemoveRangeCheck(comma, compCurStmt);
+                            // Both `tree` and `check` have been removed from the statement.
+                            // 'tree' was replaced with 'nop' or side effect list under 'comma'.
+                            return comma->gtGetOp1();
+                        }
                     }
                 }
             }
         }
-        else if (propKind == optPropKind::OPK_OBJ_GETTYPE)
-        {
-            assert(actualVal->IsCnsIntOrI());
-        }
 
 #ifdef DEBUG
         if (verbose)
         {
             printf("optEarlyProp Rewriting BB%02u\n", compCurBB->bbNum);
-            gtDispTree(root);
+            gtDispTree(compCurStmt);
             printf("\n");
         }
 #endif
-        // Rewrite the tree using a copy of "actualVal"
-        GenTreePtr actualValCopy;
-        var_types  origType = tree->gtType;
+
+        GenTreePtr actualValClone = gtCloneExpr(actualVal);
+
+        if (actualValClone->gtType != tree->gtType)
+        {
+            assert(actualValClone->gtType == TYP_LONG);
+            assert(tree->gtType == TYP_INT);
+            assert((actualConstVal >= 0) && (actualConstVal <= INT32_MAX));
```

Comment:
Previously we did check value's range for `optPropKind::OPK_ARRAYLEN`, but not for `optPropKind::OPK_OBJ_GETTYPE`,  so I added this check.